pkg/ipcache/metadata.go

```diff
@@ -0,0 +1,378 @@
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/identity/cache"
+	"github.com/cilium/cilium/pkg/labels"
+	cidrlabels "github.com/cilium/cilium/pkg/labels/cidr"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/source"
+
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// idMDMU protects the IdentityMetadata map.
+	idMDMU lock.RWMutex
+	// IdentityMetadata maps IP prefixes (x.x.x.x/32) to their labels.
+	IdentityMetadata = make(map[string]labels.Labels)
+
+	// ErrLocalIdentityAllocatorUninitialized is an error that's returned when
+	// the local identity allocator is uninitialized.
+	ErrLocalIdentityAllocatorUninitialized = errors.New("local identity allocator uninitialized")
+)
+
+// UpsertMetadata upserts a given IP and its corresponding labels associated
+// with it into the IdentityMetadata map. The given labels are not modified nor
+// is its reference saved, as their copied when inserting into the map.
+func UpsertMetadata(prefix string, lbls labels.Labels) {
+	l := labels.NewLabelsFromModel(nil)
+	l.MergeLabels(lbls)
+
+	idMDMU.Lock()
+	if cur, ok := IdentityMetadata[prefix]; !ok {
+		IdentityMetadata[prefix] = l
+	} else {
+		l.MergeLabels(cur)
+		IdentityMetadata[prefix] = l
+	}
+	idMDMU.Unlock()
+}
+
+// GetIDMetadataByIP returns the associated labels with an IP. The caller must
+// not modifying the returned object as it's a live reference to the underlying
+// map.
+func GetIDMetadataByIP(prefix string) labels.Labels {
+	idMDMU.RLock()
+	defer idMDMU.RUnlock()
+	return IdentityMetadata[prefix]
+}
+
+// InjectLabels injects labels into the IdentityMetadata (IDMD) map. The given
+// source is the source of the caller, as inserting into the IPCache requires
+// where this updated information is coming from.
+
+// Note that as this function iterates through the IDMD, if it detects a change
+// in labels for a given prefix, then this might allocate a new identity. If a
+// prefix was previously assoicated with an identity, it will get deallocated,
+// so a balance is kept.
+func InjectLabels(src source.Source, updater identityUpdater, triggerer policyTriggerer) error {
+	if !IdentityAllocator.IsLocalIdentityAllocatorInitialized() {
+		return ErrLocalIdentityAllocatorUninitialized
+	}
+
+	idMDMU.Lock()
+	defer idMDMU.Unlock()
+
+	var (
+		// trigger is true when we need to trigger policy recalculations.
+		trigger bool
+		// toUpsert stores IPKeyPairs to upsert into the ipcache.
+		toUpsert = make(map[string]Identity)
+		// idsToPropagate stores the identities that must be updated via the
+		// selector cache.
+		idsToPropagate = make(map[identity.NumericIdentity]labels.LabelArray)
+	)
+	for prefix, lbls := range IdentityMetadata {
+		id, isNew, err := injectLabels(prefix, lbls)
+		if err != nil {
+			return fmt.Errorf("failed to allocate new identity for IP %v: %w", prefix, err)
+		}
```

Comment:
Technically we could have iterated through multiple entries and successfully allocated identities for them, then say the 2nd or 3rd attempt will fail. Returning immediately here would leak the identity.

In practice, local identity allocation will not fail so I think we're OK here. We can likely simplify a bunch of the logic here by just having a consistent way to initialize all of the ipcache objects, and by removing the context / error conditions from the local identity allocation cases.